saleor/payment/gateways/np_atobarai/api.py

```diff
@@ -0,0 +1,237 @@
+    error_messages = []
+
+    if status == PaymentStatus.PENDING:
+        if cancel_error_codes := cancel(config, transaction_id).error_codes:
+            handle_unrecoverable_state(
+                order, "cancel", transaction_id, cancel_error_codes
+            )
+        error_messages = get_error_messages_from_codes(
+            action, error_codes=result["authori_hold"]
+        )
+
+    return PaymentResult(
+        status=status,
+        psp_reference=transaction_id,
+        errors=error_messages,
+    )
+
+
+def cancel_transaction(
+    config: ApiConfig, payment_information: PaymentData
+) -> PaymentResult:
+    action = TRANSACTION_CANCELLATION
+    psp_reference = payment_information.psp_reference
+
+    if not psp_reference:
+        return error_payment_result(
+            add_action_to_code(action, error_code=NO_PSP_REFERENCE)
+        )
+
+    result, error_codes = cancel(config, psp_reference)
+
+    if error_codes:
+        error_messages = get_error_messages_from_codes(action, error_codes=error_codes)
+        return errors_payment_result(error_messages)
+
+    return PaymentResult(status=PaymentStatus.SUCCESS)
+
+
+def change_transaction(
+    config: ApiConfig,
+    payment: Payment,
+    payment_information: PaymentData,
+    refund_data: Optional[Dict[int, int]],
+) -> Optional[PaymentResult]:
+    """Change transaction.
+
+    If the fulfillment was reported prior to changing given transaction,
+    then this function is a noop and return value is None.
+    """
+    if refund_data:
+        goods = get_refunded_goods(config, refund_data, payment_information)
+    else:
+        goods = get_goods_with_discount(config, payment_information)
+
+    data = {
+        "transactions": [
+            {
+                "np_transaction_id": payment.psp_reference,
+                "billed_amount": format_price(
+                    payment.captured_amount - payment_information.amount,
+                    payment_information.currency,
+                ),
+                "goods": goods,
+            }
+        ]
+    }
+
+    result, error_codes = np_request(config, "patch", "/transactions/update", json=data)
+
+    if not error_codes:
+        status = result["authori_result"]
+        transaction_id = result["np_transaction_id"]
+
+        if status == PaymentStatus.PENDING:
+            if cancel_error_codes := cancel(config, transaction_id).error_codes:
+                handle_unrecoverable_state(
+                    payment.order, "cancel", transaction_id, cancel_error_codes
+                )
+            error_messages = result["authori_hold"]
+            return errors_payment_result(error_messages)
+
+        return PaymentResult(
+            status=PaymentStatus.SUCCESS,
+        )
+
+    if PRE_FULFILLMENT_ERROR_CODE in error_codes:
+        logger.info(
+            "Fulfillment for payment with id %s was reported",
+            payment_information.graphql_payment_id,
+        )
+        return None
```

Comment:
I am not sure if I follow why we return here a `None`.  It looks like returning `None` here means that the expected case just happened. Why we don't use `PaymentResult` here? I believe we should have a separate status for handling such cases.

Reply:
Resolved in recent commit.